File: example-implementations/python/UnionFind/union_find.py
"""
Union-Find (Disjoint Set Union) data structure.
Efficiently tracks connected components with path compression and union by rank.
"""

from typing import List


class UnionFind:
    """Union-Find data structure for tracking disjoint sets."""

    def __init__(self, size: int):
        self._count = size
        self._parent: List[int] = list(range(size))
        self._rank: List[int] = [0] * size

    def count(self) -> int:
        """Return the number of disjoint sets."""
        return self._count

    def find(self, p: int) -> int:
        """Find the root of element p with path compression."""
        if self._parent[p] != p:
            self._parent[p] = self.find(self._parent[p])
        return self._parent[p]

    def union(self, p: int, q: int) -> None:
        """Unite the sets containing p and q using union by rank."""
        root_p = self.find(p)
        root_q = self.find(q)

        if root_p == root_q:
            return

        # Union by rank: attach smaller tree to larger tree
        if self._rank[root_p] < self._rank[root_q]:
            self._parent[root_p] = root_q
        elif self._rank[root_p] > self._rank[root_q]:
            self._parent[root_q] = root_p
        else:
            self._parent[root_q] = root_p
            self._rank[root_p] += 1

        self._count -= 1

    def connected(self, p: int, q: int) -> bool:
        """Check if p and q are in the same set."""
        return self.find(p) == self.find(q)

File: example-implementations/go/UnionFind/UnionFind.go
package unionfind

type UnionFind struct {
	count  int
	parent []int
	rank   []int
}

func (uf *UnionFind) Count() int { return uf.count }

func NewUnionFind(size int) *UnionFind {
	parent := make([]int, size)
	rank := make([]int, size)

	for i := 0; i < size; i++ {
		parent[i] = i
		rank[i] = 0
	}

	return &UnionFind{
		count:  size,
		parent: parent,
		rank:   rank,
	}
}

func (uf *UnionFind) Find(p int) int {
	if uf.parent[p] != p {
		uf.parent[p] = uf.Find(uf.parent[p])
	}
	return uf.parent[p]
}

func (uf *UnionFind) Union(p int, q int) {
	rootP := uf.Find(p)
	rootQ := uf.Find(q)

	if rootP == rootQ {
		return
	}

	if uf.rank[rootP] < uf.rank[rootQ] {
		uf.parent[rootP] = rootQ
	} else if uf.rank[rootP] > uf.rank[rootQ] {
		uf.parent[rootQ] = rootP
	} else {
		uf.parent[rootQ] = rootP
		uf.rank[rootP]++
	}

	uf.count--
}

func (uf *UnionFind) Connected(p int, q int) bool {
	return uf.Find(p) == uf.Find(q)
}
